sw/device/tests/sim_dv/i2c_device_tx_rx_test.c

```diff
@@ -110,47 +134,89 @@ void check_addr(uint8_t addr, dif_i2c_id_t id0, dif_i2c_id_t id1) {
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c0TxOverflow;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c0AcqFull;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c0UnexpStop;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c0HostTimeout;
+      break;
+    case 1:
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1CmdComplete;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1TxStretch;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1TxOverflow;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1AcqFull;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1UnexpStop;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c1HostTimeout;
+      break;
+    case 2:
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2CmdComplete;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2TxStretch;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2TxOverflow;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2AcqFull;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2UnexpStop;
+      plic_irqs[i++] = kTopEarlgreyPlicIrqIdI2c2HostTimeout;
+      break;
+    default:
+      LOG_FATAL("Unsupported i2c index %d", kI2cIdx);
+  }
+}
+
 bool test_main(void) {
+  LOG_INFO("Testing I2C index %d", kI2cIdx);
+
   if (kI2cByteCount > I2C_PARAM_FIFO_DEPTH - 4) {
     LOG_ERROR(
         "Test cannot fit %d bytes, 2 START records, and 2 STOP records in "
         "buffers of depth %d",
         kI2cByteCount, I2C_PARAM_FIFO_DEPTH);
   }
-  CHECK_DIF_OK(
-      dif_i2c_init(mmio_region_from_addr(TOP_EARLGREY_I2C0_BASE_ADDR), &i2c));
+
   CHECK_DIF_OK(dif_pinmux_init(
       mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+
   CHECK_DIF_OK(dif_rv_plic_init(
       mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &plic));
 
-  CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
-      &plic, kTopEarlgreyPlicIrqIdI2c0TxStretch, kTopEarlgreyPlicTargetIbex0,
-      kDifToggleEnabled));
-  CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
-      &plic, kTopEarlgreyPlicIrqIdI2c0CmdComplete, kTopEarlgreyPlicTargetIbex0,
-      kDifToggleEnabled));
-  CHECK_DIF_OK(dif_rv_plic_irq_set_priority(
-      &plic, kTopEarlgreyPlicIrqIdI2c0TxStretch, kDifRvPlicMaxPriority));
-  CHECK_DIF_OK(dif_rv_plic_irq_set_priority(
-      &plic, kTopEarlgreyPlicIrqIdI2c0CmdComplete, kDifRvPlicMaxPriority));
+  config_irqs_with_index(plic_irqs);
+  i2c_testutils_pinmux(&pinmux, kI2cIdx);
+
+  // Enable functional interrupts as well as error interrupts to make sure
+  // everything is behaving as expected.
+  for (uint32_t i = 0; i < ARRAYSIZE(plic_irqs); ++i) {
+    CHECK_DIF_OK(dif_rv_plic_irq_set_enabled(
+        &plic, plic_irqs[i], kTopEarlgreyPlicTargetIbex0, kDifToggleEnabled));
+
+    // Assign a default priority
+    CHECK_DIF_OK(dif_rv_plic_irq_set_priority(&plic, plic_irqs[i],
+                                              kDifRvPlicMaxPriority));
+  }
+
+  switch (kI2cIdx) {
+    case 0:
+      CHECK_DIF_OK(dif_i2c_init(
+          mmio_region_from_addr(TOP_EARLGREY_I2C0_BASE_ADDR), &i2c));
```

Comment:
The dif functions all use the i2c struct to pass a reference to the i2c so I'm alarmed to see it being set so late here. They might enable much more concise expressions of intent.

For example, if you use the difs you could operate on [dif_i2c_irq_t](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/dif/autogen/dif_i2c_autogen.h;bpv=1;bpt=1;l=149?gsn=dif_i2c_irq_t&gs=kythe%3A%2F%2Fgithub.com%2FlowRISC%2Fopentitan%3Flang%3Dc%252B%252B%23xscMge-OWf93wHtGpa7YEgTb3bAM2SRabz5t3uwBMAE) instead of uints and write interrupt handling code that just uses the &i2c reference to work on the correct instance.